xds/internal/balancer/edsbalancer/xds_client_wrapper.go

```diff
@@ -211,54 +252,42 @@ func (c *xdsclientWrapper) startLoadReport(edsServiceNameBeingWatched string, lo
 	}
 	c.loadReportServer = loadReportServer
 	if c.loadReportServer != nil {
-		c.cancelLoadReport = c.xdsClient.ReportLoad(*c.loadReportServer, edsServiceNameBeingWatched)
+		c.cancelLoadReport = c.xdsClient.ReportLoad(*c.loadReportServer, c.edsServiceName)
 	}
 }
 
-func (c *xdsclientWrapper) loadStore() *load.Store {
-	if c == nil || c.xdsClient == nil {
+func (c *xdsclientWrapper) loadStore() load.PerClusterReporter {
+	if c == nil || c.load.store == nil {
 		return nil
 	}
-	return c.xdsClient.LoadStore()
+	// return c.xdsClient.LoadStore().PerCluster(c.edsServiceName, "")
+	return c.load
 }
 
 // handleUpdate applies the service config and attributes updates to the client,
 // including updating the xds_client to use, and updating the EDS name to watch.
 func (c *xdsclientWrapper) handleUpdate(config *EDSConfig, attr *attributes.Attributes) {
 	clientChanged := c.updateXDSClient(config, attr)
 
-	var (
-		restartEndpointsWatch bool
-		restartLoadReport     bool
-	)
-
-	// The clusterName to watch should come from CDS response, via service
-	// config. If it's an empty string, fallback user's dial target.
-	nameToWatch := config.EDSServiceName
-	if nameToWatch == "" {
-		c.logger.Warningf("eds: cluster name to watch is an empty string. Fallback to user's dial target")
-		nameToWatch = c.bbo.Target.Endpoint
-	}
-
 	// Need to restart EDS watch when one of the following happens:
 	// - the xds_client is updated
 	// - the xds_client didn't change, but the edsServiceName changed
-	//
-	// Only need to restart load reporting when:
-	// - no need to restart EDS, but loadReportServer name changed
-	if clientChanged || c.edsServiceName != nameToWatch {
-		restartEndpointsWatch = true
-		restartLoadReport = true
-	} else if !equalStringPointers(c.loadReportServer, config.LrsLoadReportingServerName) {
-		restartLoadReport = true
-	}
-
-	if restartEndpointsWatch {
-		c.startEndpointsWatch(nameToWatch)
+	if clientChanged || c.edsServiceName != config.EDSServiceName {
```

Comment:
`c.edsServiceName` is only written to in `startEndpointsWatch`. Now that this is read from `startLoadReport`, I think it would be better to write to it here, rather than in `startEndpointsWatch`.